File: Question 0117 - Populating Next Right Pointers in Each Node II/src/Solution.java
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    private void bfs(Node root)
    {
        Queue<Node> q  = new LinkedList<>();
        q.add(root);

        while(!q.isEmpty())
        {
            int size = q.size();

            Node prev = null;
            for (int i = 0; i < size; i++) {
                Node node = q.poll();

                node.next = prev;
                prev = node;

                if(node.right != null) q.add(node.right);
                if(node.left != null) q.add(node.left);
            }
            prev = null;
        }
    }
    public Node connect(Node root)
    {
        if(root == null) return null;
        bfs(root);
        return root;
    }
}

File: Question 0117 - Populating Next Right Pointers in Each Node II/src/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Node
{
public:
    int val;
    Node *left;
    Node *right;
    Node *next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node *_left, Node *_right, Node *_next)
        : val(_val), left(_left), right(_right), next(_next) {}
};

class Solution
{
public:
    Node *connect(Node *root)
    {
        if (!root)
            return root;

        queue<Node *> q;
        q.push(root);

        while (!q.empty())
        {
            int size = q.size();

            Node *prev = NULL;

            for (int i = 0; i < size; i++)
            {
                Node *curr = q.front();
                q.pop();

                curr->next = prev;
                prev = curr;

                if (curr->right)
                    q.push(curr->right);
                if (curr->left)
                    q.push(curr->left);
            }

            prev = NULL;
        }

        return root;
    }
};